prov/sm2/src/sm2_util.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) Intel Corporation. All rights reserved.
+ * Copyright (c) Amazon.com, Inc. or its affiliates. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "config.h"
+#include "sm2_fifo.h"
+
+#include <fcntl.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+pthread_mutex_t sm2_ep_list_lock = PTHREAD_MUTEX_INITIALIZER;
+
+size_t sm2_calculate_size_offsets(ptrdiff_t *rq_offset, ptrdiff_t *fq_offset)
+{
+	size_t total_size;
+
+	/* First memory block.  The header of an sm2_region */
+	total_size = sizeof(struct sm2_region);
+
+	/* Second memory block: the recv_queue_fifo, an sm2_fifo */
+	if (rq_offset)
+		*rq_offset = total_size;
+	total_size += sizeof(struct sm2_fifo);
+
+	/* Third memory block: the message objects in a free queue */
+	if (fq_offset)
+		*fq_offset = total_size;
+	total_size += freestack_size(sizeof(struct sm2_xfer_entry),
```

Comment:
Note that if you do move to the ofi freestack instead of the smr freestack implementation, this would change since the headers are different sizes. The freestack_size function should really be smr_freestack_size

Reply:
changed.